### PR TITLE
chore: Allow Contextual Menu Component to be positioned optimally [WD-10977]

### PR DESCRIPTION
**Pull request by Kxiru**
… vertically

## Done

- Added a scss file for Contextual Menu
- Implemented initial logic by @mas-who to vertically position the Menu optimally based on the viewport height.
- Then transferred this logic from AdjustVerticalHeight function in Contextual Menu, to updateVerticalPosition function in ContextualMenuDropdown.tsx such that the menu may be adjusted according to both the container height and the viewport height.

### Storybook

To see rendered examples of all react-components, run:

```shell
yarn start
```

### QA in your project

from `react-components` run:

```shell
yarn build
npm pack
```

Install the resulting tarball in your project with:

```shell
yarn add <path-to-tarball>
```

### Percy steps

N/A

## Fixes
- Contextual Menu dropdown now displays "upwards" if there is not enough space downwards.

image

### CODE REVIEW

**Review by huwshimi — COMMENTED** [review]
It looks like this is already intended to be supported using the `autoAdjust` prop, however it currently takes into account the scroll area, not the window space.

If there's room in the scroll area below:

Screenshot 2024-06-28 at 9 28 14 AM

If there's no room below:

Screenshot 2024-06-28 at 9 28 34 AM

It looks like this could be fixed where the vertical space is calculated:

https://github.com/canonical/react-components/blob/f704e2ed5742fff7b648b9d02f64fe4d7630ad18/src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx#L228

**Comment by Kxiru** [comment]
Thanks @huwshimi for your review. Are you saying that the implemented logic would be better placed / integrated into updateVerticalPosition?

**Comment by huwshimi** [comment]
> Thanks @huwshimi for your review. Are you saying that the implemented logic would be better placed / integrated into updateVerticalPosition?

Yes please!

**Review by huwshimi — CHANGES_REQUESTED** [review]
This component already has support for adjusting the dropdown to be above or below, so there's no need to add additional state. Instead, you can just change the calculation from being relative to the scroll parent, to being relative to the window here:

https://github.com/canonical/react-components/blob/9668ce671a76d74cb3d7f1153011ad2d7c73062d/src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx#L239

**Comment by Kxiru** [comment]
> This component already has support for adjusting the dropdown to be above or below, so there's no need to add additional state. Instead, you can just change the calculation from being relative to the scroll parent, to being relative to the window here:
> 
> https://github.com/canonical/react-components/blob/9668ce671a76d74cb3d7f1153011ad2d7c73062d/src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx#L239

Don't we want the component to be able to react according to BOTH the scroll parent and the window as well?

**Comment by huwshimi** [comment]
> Don't we want the component to be able to react according to BOTH the scroll parent and the window as well?

Are you thinking of a scenario where the menu is at the bottom of a scrolling container, but that container is not at the bottom of the screen, so you want the menu to flip to above the button?

If that's the behaviour you want then you could add a second rect that is relative to the window and then adjust this calculation:

https://github.com/canonical/react-components/blob/main/src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx#L250-L252

This would use both rects to calculate "dropdown fits below in scroll parent" AND "dropdown fits below in window", in pseudo code this could be something like:

```typescript
    setVerticalPosition(
      scrollParentSpaceBelow >= dropdownHeight  && windowSpaceBelow >= dropdownHeight || windowSpaceBelow > spaceAbove ? "bottom" : "top"
    );
```

**Comment by Kxiru** [comment]
@huwshimi Hi there!
I have actioned your suggestions. Please let me know if this is in line with what you were thinking :).

File: src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx [hunk 1]
```diff
@@ -12,6 +12,7 @@ import { useWindowFitment } from "hooks";
 import Button from "../../Button";
 import type { ButtonProps } from "../../Button";
 import type { WindowFitment } from "hooks";
+// import "./ContextualMenuDropdown.scss";
```
Comment (edlerd) [inline src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx:15]:
```suggestion
```

**Review by huwshimi — COMMENTED** [review]
This is on the right track now, thanks for persisting. Right now the relation to the window isn't correct so the menu is appearing at the top when there's space below:

Screenshot 2024-07-15 at 9 38 21 AM

This PR could also use some tests, they might help during implementation to track down the issues I've mentioned.

File: src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx [hunk 2]
```diff
@@ -243,12 +244,26 @@ const ContextualMenuDropdown = <L,>({
...
+    const windowRect = {
+      top: window.scrollY,
+      bottom: window.scrollX,
+      height: window.innerHeight,
+    };
+
+    const scrollParentSpaceBelow =
+      scrollableParentRect.height - relativeRect.bottom;
+    const scrollParentSpaceAbove = relativeRect.top;
     const dropdownHeight = relativeRect.height;
```
[the remark this answers is not in the document]
Comment (Kxiru) [reply src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx:257]: Thank you for your suggestions. Please review :).

**Comment by Kxiru** [comment]
@huwshimi , thanks for our chat today! I've pushed the changes that we made and the written suggestions you provided. I'll let you know of my progress with creating a test.

**Comment by mas-who** [comment]
@huwshimi @Kxiru regarding creating a test for this component, I believe since react testing library uses jsdom, it will be tricky to setup a test that relies on the component render cycle and perform assertion on the final element positions. See this [issue](https://github.com/testing-library/react-testing-library/issues/353#issuecomment-545231016) for example which has a suggestion of directly modifying dom node's prototype to mock the values but I don't think it would work for assertions if we want to check dom element position. I think maybe we can setup the test by moving `updateVerticalPosition` outside of the component and create a unit test for the function?

**Comment by edlerd** [comment]
> @huwshimi @Kxiru regarding creating a test for this component, I believe since react testing library uses jsdom, it will be tricky to setup a test that relies on the component render cycle and perform assertion on the final element positions. See this [issue](https://github.com/testing-library/react-testing-library/issues/353#issuecomment-545231016) for example which has a suggestion to directly modify a dom node's prototype to mock the values but I don't think it would work for assertions if we want to check dom element position. I think maybe we can setup the test by moving `updateVerticalPosition` outside of the component and create a unit test for the function?

If I understand your problem correctly @mas-who @Kxiru you can have a look at the `updates the tooltip to fit on the screen` test in `Tooltip.test.jsx` for a solution of a very similar problem.

**Comment by mas-who** [comment]
> > @huwshimi @Kxiru regarding creating a test for this component, I believe since react testing library uses jsdom, it will be tricky to setup a test that relies on the component render cycle and perform assertion on the final element positions. See this [issue](https://github.com/testing-library/react-testing-library/issues/353#issuecomment-545231016) for example which has a suggestion to directly modify a dom node's prototype to mock the values but I don't think it would work for assertions if we want to check dom element position. I think maybe we can setup the test by moving `updateVerticalPosition` outside of the component and create a unit test for the function?
> 
> If I understand your problem correctly @mas-who @Kxiru you can have a look at the `updates the tooltip to fit on the screen` test in `Tooltip.test.jsx` for a solution of a very similar problem.

Hmm not sure if it's the same problem, the calculation performed for positioning the dropdown isn't really based on event listeners. As far as I can tell, what we can assert for this component test case is that it's `top` and `bottom` attributes have been correctly calculated. Note this is due to inline style being [applied](https://github.com/Kxiru/react-components/blob/a135e875be8ec29d05723d20ce62f911c0d3afa9/src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx#L311). To do that, we need to get the dropdown element from the dom after rendering the component, which would have zero dimensions as discussed in the reference issue. This would cause assertion to fail.

**Comment by huwshimi** [comment]
To get this to work you'll need to mock the functions that don't have dimensions, like we do in this test: https://github.com/canonical/react-components/blob/2825eb98ed253825ef69e63c261c7c9605b15f4f/src/hooks/useWindowFitment.test.ts#L72-L78.

You'll need to do this for the scroll parent, the position node and the dropdown elements.

You'll also need to double render the component as RTL doesn't do full lifecycle rendering so we have to manually rerender the component to make the hooks run after the elements exist in the DOM:

https://github.com/canonical/react-components/blob/2825eb98ed253825ef69e63c261c7c9605b15f4f/src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.test.tsx#L42-L44.

If that doesn't make sense let me know and I can give you a concrete example of how to construct the test.

**Comment by Kxiru** [comment]
@huwshimi , @mas-who has pushed a commit that  recreates the test environment to check whether VerticalPosition resolves to top or bottom - and consequently, whether the styling attribute of the component Span has a bottom of 0, or not. Please could you let us know your thoughts of the test implementation?

File: src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx [hunk 3]
```diff
@@ -243,27 +243,19 @@ const ContextualMenuDropdown = <L,>({
       bottom: toggleRect.bottom - scrollableParentRect.top,
     };
 
-    // Calculate the rect in relation to the Window
-    const windowRect = {
```
Comment (Kxiru) [inline src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.tsx]: For context, this has been removed as the only attribute that is later referenced, is windowRect.height, i.e; window.innerHeight - as replaced/referenced in line 252.

**Comment by huwshimi** [comment]
> @huwshimi , @mas-who has pushed a commit that recreates the test environment to check whether VerticalPosition resolves to top or bottom - and consequently, whether the styling attribute of the component Span has a bottom of 0, or not. Please could you let us know your thoughts of the test implementation?

This looks good! I think there are two other cases that need tests:
- that the menu is above if there is enough area in the window, but not in the scroll parent
- that the menu is below if there is not enough area in the window or scroll parent, but also that there is more space below than above.

I also think it could be worth splitting the current test into the two assertions to make it clear which mocks are required to make each test pass. It might be helpful to move some of the set up into a function to make the four tests cleaner.

**Comment by mas-who** [comment]
> > @huwshimi , @mas-who has pushed a commit that recreates the test environment to check whether VerticalPosition resolves to top or bottom - and consequently, whether the styling attribute of the component Span has a bottom of 0, or not. Please could you let us know your thoughts of the test implementation?
> 
> This looks good! I think there are two other cases that need tests:
> 
> * that the menu is above if there is enough area in the window, but not in the scroll parent
> * that the menu is below if there is not enough area in the window or scroll parent, but also that there is more space below than above.
> 
> I also think it could be worth splitting the current test into the two assertions to make it clear which mocks are required to make each test pass. It might be helpful to move some of the set up into a function to make the four tests cleaner.

@huwshimi I've made those changes as mentioned :+1:

File: src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.test.tsx [hunk 4]
```diff
@@ -171,4 +171,180 @@ describe("ContextualMenuDropdown ", () => {
       ).toBe("right");
     });
   });
+
+  describe("dropdown menu vertical positioning", () => {
+    const renderMenu = (positionNode: HTMLElement, menuHeight: number) => {
+      // need to mock overflow propertty for scrollparent
```
Comment (huwshimi) [inline src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.test.tsx:177]:
```suggestion
      // need to mock overflow property for scrollparent
```

**Review by huwshimi — APPROVED** [review]
Thanks for this @Kxiru and @mas-who this looks like it's working well now!

File: src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.test.tsx (hunk 4 above)
Comment (mas-who) [reply src/components/ContextualMenu/ContextualMenuDropdown/ContextualMenuDropdown.test.tsx]: Thanks for the review @huwshimi !

**Review by edlerd — APPROVED** [review]
LGTM

**Pull request merged**